### PR TITLE
Refactor editor.js with Object-Oriented Principles

### PR DESCRIPTION
**Pull request by noi5e**
See #9004

I think that went a lot easier than I expected! This seems to work okay testing in the local environment, lets see how the system tests do. Will update this comment if there's anything significant.

**UPDATE:** 

Ready to merge! 🙌 

I know this looks like a lot of commits, but most of them were just to pass codeclimate. Is it just me or did it get really strict all of a sudden?

---
(This issue is part of the larger Comment Editor Overhaul Project with Outreachy. Refer to Planning Issue #9069 for more context)

### CODE REVIEW

File: app/assets/javascripts/editor.js [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
-    // default parameters point toward either:
-    //   1. the comment form at the bottom of multi-comment wikis/questions/research notes
-    //   2. the only editor form on /wiki/new and /wiki/edit
-    $E.setState();
-    
+class Editor {
+  // default parameters reference the IDs of:
+  //   1. main comment form in multi-comment wikis, questions, & research notes.
+  //   2. the only editor form on /wiki/new and /wiki/edit
+  constructor(textarea = "text-input", preview = "comment-preview-main", title = "title") {
```
Comment (noi5e) [inline app/assets/javascripts/editor.js:5]: These default params are just a stepping stone for now. 

Since the element IDs are all standardized now, it's easier to do a rewrite so that constructor takes just **one** parameter. For example: 

`$E = new Editor('reply-1234');` means that the editor object has access to both `#comment-preview-**reply-1234**` and `#text-input-**reply-1234**`. 

Much simpler.

File: app/assets/javascripts/editor.js [hunk 2]
```diff
@@ -23,36 +31,36 @@ $E = {
...
-  },
-  is_editing: function() {
-    return ($E.textarea[0].selectionStart == 0 && $E.textarea[0].selectionEnd == 0)
-  },
-  refresh: function() {
+  };
+  // code seems unused, commenting out for now.
+  // is_editing = function() {
+  //   return ($E.textarea[0].selectionStart == 0 && $E.textarea[0].selectionEnd == 0)
+  // };
```
Comment (noi5e) [inline app/assets/javascripts/editor.js:44]: Couldn't find this code used anywhere doing a search for `is_editing` in VSCode. Will delete in the next few PRs if the tests keep passing.

File: app/assets/javascripts/editor.js [hunk 3]
```diff
@@ -23,36 +31,36 @@ $E = {
         return code;
       }
     });
-  },
-  setState: function(textarea = 'text-input', preview = 'comment-preview-main', title = 'title') {
+  }
+  setState = function(textarea = 'text-input', preview = 'comment-preview-main', title = 'title') {
```
[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js:35]: I'm a little stumped about this error here. Maybe the `constructor` function before it needs a semicolon after the curly brace, like `};`? Will try that in the morning.

File: app/assets/javascripts/editor.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js:5]: I think I will have to delete templates to get this PR to pass this check.

File: app/assets/javascripts/editor.js [hunk 4]
```diff
@@ -1,13 +1,21 @@
...
-
-$E = {
-  initialize: function() {
-    // call setState with no parameters, aka. default parameters.
-    // default parameters point toward either:
-    //   1. the comment form at the bottom of multi-comment wikis/questions/research notes
-    //   2. the only editor form on /wiki/new and /wiki/edit
-    $E.setState();
-    
+class Editor {
```
[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js:1]: I could get around this by commenting out the `h1`, and `h3` - `h7` methods that aren't currently used anywhere in production.

On the other hand, I could also delete template functionality, which is on my to-dos.

File: app/assets/javascripts/editor.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js]: Figured it out... In OOP JS, object methods are written like `setState()` instead of `setState = function()`.

File: app/assets/javascripts/editor.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js:5]: Just decided to remove `\n` in the templates definition (it's going to get deleted anyway) to get this to pass.

File: app/assets/javascripts/editor.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js:1]: Commented out the header methods to pass this. I have a bad habit of making PRs too big and out-of-scope.

File: app/assets/javascripts/editor.js [hunk 5]
```diff
@@ -16,43 +19,43 @@ $E = {
...
-  },
+    const singleFormPath = RegExp(/\/(wiki|features)(\/[^\/]+\/edit|\/new)/);
+    return singleFormPath.test(url);
+  }
   // wraps currently selected text in textarea with strings a and b
-  wrap: function(a, b, args) {
-    // we only refresh $E's values if we are on a page using the rich-text editor (most pages).
-    // the legacy editor pages only have one editor form, unlike pages with multiple comments.
-    if (this.isRichTextEditor(window.location.pathname)) { this.refresh(); }
+  wrap(a, b, args) {
```
[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js:56]: Yikes, I remember codeclimate was giving me a lot of grief [in this PR](https://github.com/publiclab/plots2/pull/9035#discussion_r559910489) as well.

I did a little bit of work in the `wrap` method (mostly renaming methods and writing new comments), which started triggering this error. Then I completely scaled it back, removing all the changes, and still got the error. Weird.

[the remark this answers is not in the document]
Comment (noi5e) [reply app/assets/javascripts/editor.js]: This is passing now.

File: app/assets/javascripts/editor.js (hunk 1 above)
Comment (Sagarpreet) [reply app/assets/javascripts/editor.js:5]: Nice !!

File: app/assets/javascripts/editor.js [hunk 6]
```diff
@@ -23,117 +26,110 @@ $E = {
...
-    if (newlineParameterExists) {
-      replace = replace + "\n\n";
-    }
-    if (newlineParameterExists && $E.textarea[0].selectionStart > 0) { 
-      replace = "\n" + replace; 
-    }
-    $E.textarea.val($E.textarea.val().substring(0,start) + replace + $E.textarea.val().substring(end,len));
-  },
-  bold: function() {
+  wrap(a, b, newlineDesired = false, fallback) {
```
Comment (Sagarpreet) [inline app/assets/javascripts/editor.js:56]: can we name it better than `a` and `b` for new contributors to understand this better?

File: app/assets/javascripts/editor.js (hunk 2 above)
Comment (Sagarpreet) [reply app/assets/javascripts/editor.js:39]: Let's remove and clean this then 💯

File: app/assets/javascripts/editor.js [hunk 7]
```diff
@@ -23,117 +26,110 @@ $E = {
...
+    // we only refresh $E's values if we are on a page with multiple comments
+    if (!this.isSingleFormPage(window.location.pathname)) { this.refresh(); }
+
+    const selectionStart = $E.textarea[0].selectionStart;
+    const selectionEnd = $E.textarea[0].selectionEnd;
+    const selection = fallback || $E.textarea.val().substring(selectionStart, selectionEnd); // fallback if nothing has been selected, and we're simply dealing with an insertion point
+
+    let newText = a + selection + b; // ie. ** + selection + ** (wrapping selection in bold)
+    if (newlineDesired) { newText = newText + "\n\n"; }
```
Comment (Sagarpreet) [inline app/assets/javascripts/editor.js:65]:
```suggestion
    let newText = newlineDesired ?  a + selection + b + "\n\n" : a + selection + b; // ie. ** + selection + ** (wrapping selection in bold)
```

File: app/assets/javascripts/editor.js [hunk 8]
```diff
@@ -23,117 +26,110 @@ $E = {
...
-  italic: function() {
+  }
+  italic() {
     $E.wrap('_','_')
-  },
-  link: function(uri) {
+  }
+  link(uri) {
     uri = prompt('Enter a URL');
     if (uri === null) { uri = ""; }
```
Comment (Sagarpreet) [inline app/assets/javascripts/editor.js:82]:
```suggestion
    if (!!uri) { uri = ""; }
```

File: app/assets/javascripts/editorToolbar.js [hunk 9]
```diff
@@ -108,21 +108,17 @@ $(function() {
...
               orig_image_url = file_url + '?s=o' // size = original
-              $E.wrap('[![', '](' + file_url + ')](' + orig_image_url + ')', {'newline': true, 'fallback': data.result['filename']}) // on its own line; see /app/assets/js/editor.js
+              $E.wrap('[![', '](' + file_url + ')](' + orig_image_url + ')', true, data.result['filename']);
               break;
             case 'csv':
-              $E.wrap('[graph:' + file_url + ']', '', {'newline': true})
+              $E.wrap('[graph:' + file_url + ']', '', true);
               break;
             default:
-              $E.wrap('<a href="'+data.result.url.split('?')[0]+'"><i class="fa fa-file"></i> ','</a>', {'newline': true, 'fallback': data.result['filename'].replace(/[()]/g , "-")}) // on its own line; see /app/assets/js/editor.js
```
Comment (Sagarpreet) [inline app/assets/javascripts/editorToolbar.js:117]: Why was fallback used? and why is it removed now?

File: app/assets/javascripts/post.js [hunk 10]
```diff
@@ -2,7 +2,7 @@ jQuery(document).ready(function() {
 
   $('.datepicker').datepicker()
   
-  $E.initialize()
+  $E = new Editor();
```
Comment (Sagarpreet) [inline app/assets/javascripts/post.js:5]: 🎉

File: app/views/questions/show.html.erb [hunk 11]
```diff
@@ -1,10 +1,14 @@
 <%= javascript_include_tag('comment_expand') %>
 <%= javascript_include_tag('notes') %>
 <%= javascript_include_tag('textbox_expand') %>
-<script> var comments_length = <%= @node.comments.length %>; $(function(){
-  $E.initialize();
-  $("img").lazyload();
-});</script>
+<script> 
+  var comments_length = <%= @node.comments.length %>;
```
Comment (Sagarpreet) [inline app/views/questions/show.html.erb:5]: This is not getting used anywhere, right?

File: app/assets/javascripts/editor.js (hunk 6 above)
Comment (noi5e) [reply app/assets/javascripts/editor.js:56]: I totally agree about renaming `a` and `b`!!! I'll do it soon.

File: app/assets/javascripts/editor.js (hunk 2 above)
Comment (noi5e) [reply app/assets/javascripts/editor.js:39]: Deleted it in another PR!

File: app/assets/javascripts/editor.js (hunk 7 above)
Comment (noi5e) [reply app/assets/javascripts/editor.js:65]: Thank you!!!!!

File: app/assets/javascripts/editorToolbar.js (hunk 9 above)
Comment (noi5e) [reply app/assets/javascripts/editorToolbar.js:117]: This is the only place where `fallback` is still used (3 times):
https://github.com/publiclab/plots2/blob/97247e30289dfb27f93446b84d036ad9811e7983/app/assets/javascripts/editorToolbar.js#L108-L118

I didn't delete `fallback` here, but I just separated it out into two parameters instead of one object:
- **Before**: `$E.wrap(a, b, { fallback: "fallback", newline: "true" });`
- **After**: `$E.wrap(a, b, true, "fallback")`;`

In other words, `wrap` now takes four parameters, instead of three. And none of the parameters are objects.

I know it might be a little harder to read but I really needed to do this to pass codeclimate. It was failing this PR saying that the old code in `$E.wrap` had too much 'cognitive complexity'. Part of the reason why was because the parameter was an object, and we had to do a check like `const fallBackExists = args && args['fallback']`&mdash;which adds to cognitive complexity.

**tldr;** I didn't delete fallback, the same functionality is still there. Just changed parameters.

File: app/views/questions/show.html.erb (hunk 11 above)
Comment (noi5e) [reply app/views/questions/show.html.erb:5]: `comments_length`? I think it still is here:
https://github.com/publiclab/plots2/blob/97247e30289dfb27f93446b84d036ad9811e7983/app/assets/javascripts/question.js#L13-L23

Comment (noi5e) [reply app/views/questions/show.html.erb:5]: But I don't know what this code does 😅

File: app/assets/javascripts/editor.js (hunk 8 above)
Comment (noi5e) [reply app/assets/javascripts/editor.js:81]: Interesting, I've never heard of `!!` to check for `null`!

Comment (sagarpreet-chadha) [reply app/assets/javascripts/editor.js:81]: Hey it is single !

Comment (sagarpreet-chadha) [reply app/assets/javascripts/editor.js:81]: So !! gives false for undefined, null, 0 and “”

Comment (noi5e) [reply app/assets/javascripts/editor.js:81]: Okay! I think it makes sense to check for `undefined` too, so will commit this. Thank you!

**Comment by noi5e** [comment]
@jywarren @Sagarpreet ready to merge again!

**Review by sagarpreet-chadha — APPROVED** [review]

File: app/assets/javascripts/editor.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jywarren) [reply app/assets/javascripts/editor.js]: Great! We don't ALWAYS have to agree with CC on every suggestion. But i think long term it makes for better code, so I really appreciate it!

File: app/assets/javascripts/post.js (hunk 10 above)
Comment (jywarren) [reply app/assets/javascripts/post.js:5]: 🎉 🎉 🎉  niceeeeee

File: app/views/questions/show.html.erb (hunk 11 above)
Comment (jywarren) [reply app/views/questions/show.html.erb:5]: This looks like old code related to the "Answers" system which we've deprecated; answers were back turned into comments. It seems it was for showing only the first... 3 answers, hiding the rest behind a "show more" kind of thing? I think that's all gone now, so we should be good to remove it.

**Review by jywarren — APPROVED** [review]
> I know this looks like a lot of commits, but most of them were just to pass codeclimate. Is it just me or did it get really strict all of a sudden?

Hmm. i wonder if this is just an old and messy code file and we had previously dismissed CC suggestions, but re-editing the file is bringing them all back up?

This looks super though, thank you! Let's remember to circle back to other thoughts here like removing segments, in a follow-up PR; @noi5e if you can drop them into an issue for now that'd be great!
